posthog/heatmaps/sql.py

```diff
@@ -0,0 +1,192 @@
+    scale_factor Int16,
+    viewport_width Int16,
+    viewport_height Int16,
+    -- some elements move when the page scrolls, others do not
+    pointer_target_fixed Bool,
+    current_url VARCHAR,
+    type LowCardinality(String)
+) ENGINE = {engine}
+"""
+
+HEATMAPS_TABLE_BASE_SQL = """
+CREATE TABLE IF NOT EXISTS {table_name} ON CLUSTER '{cluster}'
+(
+    session_id VARCHAR,
+    team_id Int64,
+    distinct_id VARCHAR,
+    timestamp DateTime64(6, 'UTC'),
+    -- x is the x with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
+    x Int16,
+    -- y is the y with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
+    y Int16,
+    -- stored so that in future we can support other resolutions
+    scale_factor Int16,
+    viewport_width Int16,
+    viewport_height Int16,
+    -- some elements move when the page scrolls, others do not
+    pointer_target_fixed Bool,
+    current_url VARCHAR,
+    type LowCardinality(String),
+    _timestamp DateTime
+) ENGINE = {engine}
+"""
+
+HEATMAPS_DATA_TABLE_ENGINE = lambda: MergeTreeEngine("heatmaps", replication_scheme=ReplicationScheme.SHARDED)
+
+HEATMAPS_TABLE_SQL = lambda: (
+    HEATMAPS_TABLE_BASE_SQL
+    + """
+    PARTITION BY toYYYYMM(timestamp)
+    -- almost always this is being queried by
+    --   * type,
+    --   * team_id,
+    --   * date range,
+    --   * URL (maybe matching wild cards),
+    --   * width
+    -- we'll almost never query this by session id
+    -- so from least to most cardinality that's
+    ORDER BY (type, team_id,  toDate(timestamp), current_url, viewport_width)
+-- I am purposefully not setting index granularity
+-- the default is 8192, and we will be loading a lot of data
+-- per query, we tend to copy this 512 around the place but
+-- i don't think it applies here
+"""
+).format(
+    table_name=HEATMAPS_DATA_TABLE(),
+    cluster=settings.CLICKHOUSE_CLUSTER,
+    engine=HEATMAPS_DATA_TABLE_ENGINE(),
+)
+
+KAFKA_HEATMAPS_TABLE_SQL = lambda: KAFKA_HEATMAPS_TABLE_BASE_SQL.format(
+    table_name="kafka_heatmaps",
+    cluster=settings.CLICKHOUSE_CLUSTER,
+    engine=kafka_engine(topic=KAFKA_CLICKHOUSE_HEATMAP_EVENTS),
+)
+
+HEATMAPS_TABLE_MV_SQL = (
+    lambda: """
+CREATE MATERIALIZED VIEW IF NOT EXISTS heatmaps_mv ON CLUSTER '{cluster}'
+TO {database}.{target_table}
+AS SELECT
+    session_id,
+    team_id,
+    distinct_id,
+    timestamp,
+    -- x is the x with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
+    x,
+    -- y is the y with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
+    y,
+    -- stored so that in future we can support other resolutions
+    scale_factor,
+    viewport_width,
+    viewport_height,
+    -- some elements move when the page scrolls, others do not
+    pointer_target_fixed,
+    current_url,
+    type,
+    _timestamp
```

Comment:
same thing here:
something else that would be lovely to add here are the following [virtual columns](https://clickhouse.com/docs/en/engines/table-engines/integrations/kafka#virtual-columns):
_offset
_partition

We've really wished to have these on tables in the past - would be super cool to start adding them here.